File: app/src/main/java/com/recursia/popularmovies/domain/DetailScreenInteractorImpl.java
package com.recursia.popularmovies.domain;

import com.recursia.popularmovies.domain.models.Movie;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;

public class DetailScreenInteractorImpl implements DetailScreenInteractor {

    private final MoviesRepository moviesRepository;

    public DetailScreenInteractorImpl(MoviesRepository moviesRepository) {
        this.moviesRepository = moviesRepository;
    }

    @Override
    public Single<List<Movie>> getMovieRecommendations(int movieId, int page, String language) {
        return moviesRepository.getMovieRecommendations(movieId, page, language);
    }

    @Override
    public Single<Movie> getMovieById(int movieId, String language) {
        return moviesRepository.getMovieById(movieId, language);
    }

    @Override
    public Completable makeFavoriteMovie(Movie movie) {
        return moviesRepository.makeFavoriteMovie(movie);
    }

    @Override
    public Completable removeFavoriteMovie(Movie movie) {
        return moviesRepository.removeFavoriteMovie(movie);
    }

}

File: app/src/main/java/com/recursia/popularmovies/domain/DetailScreenInteractorImpl.kt
package com.recursia.popularmovies.domain

import com.recursia.popularmovies.domain.models.Movie
import com.recursia.popularmovies.domain.models.Review
import io.reactivex.Completable
import io.reactivex.Single

class DetailScreenInteractorImpl(
        private val accountRepository: AccountRepository,
        private val moviesRepository: MoviesRepository
) : DetailScreenInteractor {
    override fun getMovieById(movieId: Int, language: String): Single<Movie> {
        return accountRepository.getMovieById(movieId)
                .switchIfEmpty(moviesRepository.getMovieById(movieId, language))
    }

    override fun setMovieStatus(movie: Movie): Completable {
        return accountRepository.setMovieStatus(movie)
    }

    override fun getMovieRecommendations(movieId: Int, page: Int, language: String): Single<List<Movie>> {
        return moviesRepository.getMovieRecommendations(movieId, page, language)
    }

    override fun translateReview(review: Review, lang: String): Single<Review> {
        TODO("Not yet implemented")
    }
}
